generated/nifake/nifake/_grpc.py

```diff
@@ -0,0 +1,529 @@
+# -*- coding: utf-8 -*-
+# This file was generated
+
+import grpc
+import hightime  # noqa: F401
+import warnings
+
+from . import enums as enums
+from . import errors as errors
+from . import nifake_pb2 as grpc_types
+from . import nifake_pb2_grpc as grpc_library
+
+from . import custom_struct as custom_struct  # noqa: F401
+
+from . import custom_struct_nested_typedef as custom_struct_nested_typedef  # noqa: F401
+
+from . import custom_struct_typedef as custom_struct_typedef  # noqa: F401
+
+
+class LibraryInterpreter(object):
+    '''LibraryInterpreter
+
+    Wrapper around driver grpc library.
+    '''
+
+    def __init__(self, grpc_channel):
+        self._client = grpc_library.NiFakeStub(grpc_channel)
+        self._vi = 0
+
+    def _invoke(self, func, request):
+        grpc_error = None
+        try:
+            response = func(request)
+            error_code = response.status
+            error_message = ''
+        except grpc.RpcError as rpc_error:
+            error_code = None
+            error_message = rpc_error.details()
+            for entry in rpc_error.trailing_metadata() or []:
+                if entry.key == 'ni-error':
+                    value = entry.value if isinstance(entry.value, str) else entry.value.decode('utf-8')
+                    try:
+                        error_code = int(value)
+                    except ValueError:
+                        error_message += f'\nError status: {value}'
+            grpc_error = rpc_error.code()
+
+        if grpc_error == grpc.StatusCode.UNAVAILABLE:
+            error_message = 'Failed to connect to server'
+        elif grpc_error == grpc.StatusCode.UNIMPLEMENTED:
+            error_message = (
+                'The operation is not implemented or is not supported/enabled in this service'
+            )
+        elif grpc_error == grpc.StatusCode.NOT_FOUND:
+            raise errors.DriverTooOldError()
+
+        if grpc_error == grpc.StatusCode.INVALID_ARGUMENT:
+            raise ValueError(error_message)
+        elif error_code is None:
+            raise errors.Error(error_message)
+        elif error_code < 0:
+            raise errors.DriverError(error_code, error_message)
+        elif error_code > 0:
+            if not error_message:
+                error_message = self.get_error_description(error_code)
+            warnings.warn(errors.DriverWarning(error_code, error_message))
+        return response
+
+    def get_error_description(self, error_code):
```

Comment:
This method makes no sense in gRPC. In the warning case above, you should call `self.error_message(error_code)` directly.

Why? Once the session is valid, calling the get_error is thread-specific. Since the thread it will execute on is random, it will not return anything useful.

The message below about IVI requiring GetError to fail when the session is invalid feels totally wrong, by the way, even in the library interpreter. It returns the last error from the current thread if the session is NULL but returns the last error on the current session from any thread if the session is not NULL.

Reply:
Done (removed `get_error_description`; using `error_message` instead).